sdks/java/io/redis/src/test/java/org/apache/beam/sdk/io/redis/RedisIOTest.java

```diff
@@ -0,0 +1,186 @@
+            return service;
+          }
+        });
+    long estimatedSizeBytes = source.getEstimatedSizeBytes(pipelineOptions);
+    LOG.info("Estimated size: {}", estimatedSizeBytes);
+    assertEquals(100, estimatedSizeBytes);
+  }
+
+  @Test
+  @Category(RunnableOnService.class)
+  public void testGetRead() throws Exception {
+    FakeRedisService service = new FakeRedisService();
+    for (int i = 0; i < 1000; i++) {
+      service.put("Key " + i, "Value " + i);
+    }
+
+    PCollection<KV<String, String>> output =
+        pipeline.apply(RedisIO.read().withRedisService(service));
+
+    PAssert.thatSingleton(output.apply("Count", Count.<KV<String, String>>globally()))
+        .isEqualTo(1000L);
+    ArrayList<KV<String, String>> expected = new ArrayList<>();
+    for (int i = 0; i < 1000; i++) {
+      KV<String, String> kv = KV.of("Key " + i, "Value " + i);
+      expected.add(kv);
+    }
+    PAssert.that(output).containsInAnyOrder(expected);
+
+    pipeline.run();
+  }
+
+  /**
+   * A {@link RedisService} implementation that stores key/value pairs in memory.
+   */
+  private static class FakeRedisService implements RedisService {
+
+    private static final Map<String, String> store = new HashMap<>();
+
+    public void put(String key, String value) {
+      store.put(key, value);
+    }
+
+    public String get(String key) {
+      return store.get(key);
+    }
+
+    public void clear() {
+      store.clear();
+    }
+
+    @Override
+    public FakeRedisReader createReader(String keyPattern) throws IOException {
+      return new FakeRedisReader(keyPattern);
+    }
+
+    static class FakeRedisReader implements Reader {
+
+      private final String keyPattern;
+      private Iterator<String> keysIterator;
+      private KV<String, String> current;
+
+      public FakeRedisReader(String keyPattern) {
+        this.keyPattern = keyPattern;
+      }
+
+      @Override
+      public boolean start() {
+        keysIterator = store.keySet().iterator();
+        return advance();
+      }
+
+      @Override
+      public boolean advance() {
+        if (keysIterator.hasNext()) {
+          String key = keysIterator.next();
+          if (keyPattern != null && !keyPattern.equals("*")) {
+            if (!key.matches(keyPattern)) {
+              return false;
+            }
+          }
+          String value = store.get(key);
+          KV<String, String> record = KV.of(key, value);
+          current = record;
+          return true;
+        }
+        return false;
+      }
+
+      @Override
+      public void close() {
+        keysIterator = null;
+        current = null;
+      }
+
+      @Override
+      public KV<String, String> getCurrent() {
+        if (current == null) {
+          throw new NoSuchElementException();
+        }
+        return current;
+      }
+
+    }
+
+    @Override
+    public long getEstimatedSizeBytes() {
+      long size = 0L;
+      Iterator<String> keysIterator = store.keySet().iterator();
```

Comment:
You can iterate over it using a regular foreach loop. Also why do you need to look only at the 10 first elements in this mock implementation?